content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+put your energy into the things that really interest you.
+
+# Compiler Explorer
+
+Of course we have TableGen in Compiler Explorer! Is a language even real if it’s
+not in Compiler Explorer?
+
+(Of course it is, but if you’re favourite language isn’t, Compiler Explorer has
+[excellent documentation](https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingALanguage.md)
+and friendly maintainers)
+
+Compiler Explorer is a whole bunch of different versions of compilers for
+different languages and different architectures that you can access with just a
+browser tab.
+
+It’s an incredible tool for learning, teaching, triaging, optimising and
+[many more](https://www.youtube.com/watch?v=O5sEug_iaf4) things. I won’t go into
+detail about it here, just a few things about TableGen’s inclusion.
+
+The obvious thing is that `llvm-tblgen` doesn’t emit machine code (though a
+hypothetical backend could) so there is no option to compile to binary or
+execute code.
+
+By default you will see the records printed as plain text. A backend can be
+chosen by adding a compiler option manually or as an “Action” from the
+“Overrides” menu.
+
+It’s important to note that TableGen backends have very specific expectations of
+what will be in the source code. Almost as if you had a C++ compiler that
+wouldn’t compile for Arm unless it saw the symbol `arm_is_cool` somewhere in the
+source code.
+
+In the LLVM monorepo all the required classes are set up for you, but in
+Compiler Explorer they are not. So if you want to experiment with an existing
+backend I suggest providing stub implementations of the classes, or copying some
+in from the LLVM project. Standard includes from `include/llvm/*.td` can also be
+used.
+
+Developing a backend within Compiler Explorer is not possible at this time, but
+you can select the JSON backend and copy that JSON to give to local scripts.
+
+Multi-file projects (“IDE mode”) also work as expected, so you can have your own
+[include files](https://godbolt.org/z/4qhdoaMjE) if you like.
+
+Finally, remember that Compiler Explorer examples can be shared. If you’re
+asking or answering questions about TableGen, always include a Compiler Explorer
+link if you can!
+
+# Jupyter Notebooks
+
+[Jupyter](https://jupyter.org/) creates interactive notebooks that combine text
+and code into one document. With the ability to edit that code rerun it to
+update the results in the notebook.
+
+This is great for taking notes or building up large examples from small chunks
+of code. Then you can export your notes as a notebook anyone can edit, or in
+static formats like PDF or Markdown.
+
+TableGen can be used in one of these notebooks via the
+[TableGen Jupyter Kernel](https://github.com/llvm/llvm-project/tree/main/llvm/utils/TableGen/jupyter).
+Install instructions are on the linked page and you can watch me talk more about
+it [here](https://www.youtube.com/watch?v=Gf0FUiY2TRo).
```

Comment:
Installation instructions are available here[link page] and you can watch me talk more.....